src/query/sql/src/planner/optimizer/optimizer.rs

```diff
@@ -480,32 +480,19 @@ async fn optimize_merge_into(mut opt_ctx: OptimizerContext, s_expr: SExpr) -> Re
         })
     }
 
-    plan.change_join_order = join_order_changed;
-    if opt_ctx.enable_distributed_optimization {
-        let merge_source_optimizer = MergeSourceOptimizer::create();
-        // Inner join shouldn't add `RowNumber` node.
-        let mut enable_right_broadcast = false;
-
-        if matches!(join_op.join_type, JoinType::RightAnti | JoinType::Right)
-            && merge_source_optimizer
-                .merge_source_matcher
-                .matches(&join_s_expr)
+    let distributed = !join_s_expr.has_merge_exchange();
+    if opt_ctx.enable_distributed_optimization && distributed {
+        let merge_optimizer = MergeOptimizer::create();
+        // now always use shuffle
+        if matches!(
+            join_op.join_type,
+            JoinType::Left | JoinType::LeftAnti | JoinType::Inner
+        ) && merge_optimizer.merge_matcher.matches(&join_s_expr)
```

Comment:
Why non-right join to enforce shuffle join?

We should respect the join distribution type(shuffle or broadcast).

Reply:
Yes, if the join type is inner join, we can use broadcast join if the source table is on the build side.